File: web/src/main/java/hu/bsstudio/safetycar/configuration/SafetyCarRouterConfiguration.java
package hu.bsstudio.safetycar.configuration;

import hu.bsstudio.safetycar.SafetyCarFollowHandler;
import hu.bsstudio.safetycar.SafetyCarOvertakeHandler;
import hu.bsstudio.safetycar.SafetyCarService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.RouterFunctions;
import org.springframework.web.reactive.function.server.ServerResponse;

@Configuration
public class SafetyCarRouterConfiguration {

    @Autowired
    private SafetyCarService service;

    @Bean
    public RouterFunction<ServerResponse> safetyCarRouterFunction(final SafetyCarFollowHandler safetyCarFollowHandler,
                                                                  final SafetyCarOvertakeHandler safetyCarOvertakeHandler) {
        return RouterFunctions.route()
            .POST("/api/speed/safetyCar/follow", safetyCarFollowHandler)
            .POST("/api/speed/safetyCar/overtake", safetyCarOvertakeHandler)
            .build();
    }

    @Bean
    public SafetyCarFollowHandler safetyCarFollowHandler() {
        return new SafetyCarFollowHandler(service);
    }

    @Bean
    public SafetyCarOvertakeHandler safetyCarOvertakeHandler() {
        return new SafetyCarOvertakeHandler(service);
    }
}

File: server/web/src/main/kotlin/hu/bsstudio/robonaut/safetycar/configuration/SafetyCarRouterConfiguration.kt
package hu.bsstudio.robonaut.safetycar.configuration

import hu.bsstudio.robonaut.safetycar.SafetyCarFollowHandler
import hu.bsstudio.robonaut.safetycar.SafetyCarOvertakeHandler
import hu.bsstudio.robonaut.safetycar.SafetyCarService
import hu.bsstudio.robonaut.security.RobonAuthFilter
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.reactive.function.server.RouterFunction
import org.springframework.web.reactive.function.server.RouterFunctions
import org.springframework.web.reactive.function.server.ServerResponse

@Configuration
class SafetyCarRouterConfiguration(
  private val robonAuthFilter: RobonAuthFilter,
  private val safetyCarService: SafetyCarService,
) {
  @Bean
  fun safetyCarRouterFunction(
    safetyCarFollowHandler: SafetyCarFollowHandler,
    safetyCarOvertakeHandler: SafetyCarOvertakeHandler,
  ): RouterFunction<ServerResponse> =
    RouterFunctions
      .route()
      .filter(robonAuthFilter)
      .path("/api/speed/safetyCar") { builder ->
        builder
          .POST("/follow", safetyCarFollowHandler)
          .POST("/overtake", safetyCarOvertakeHandler)
          .build()
      }.build()

  @Bean
  fun safetyCarFollowHandler() = SafetyCarFollowHandler(safetyCarService)

  @Bean
  fun safetyCarOvertakeHandler() = SafetyCarOvertakeHandler(safetyCarService)
}
